### PR TITLE
Artnet Demux

### PR DESCRIPTION
**Pull request by meawoppl**
Add the demuxing output.

### CODE REVIEW

**Comment by Geethree** [comment]
Hailing @mcslee and @chaqke to checkout our demuxing in https://github.com/temple2018/Galaxia/blob/a470d06ba6264257fa5c8804cfc5385e21050e89/src/main/java/org/templegalaxia/DemuxArtNetDatagram.java

File: build.gradle [hunk 1]
```diff
@@ -1,7 +1,24 @@
+plugins {
+    id "com.diffplug.gradle.spotless" version "3.13.0"
+}
+
 apply plugin: 'application'
 apply plugin: 'java'
 
 
+spotless {
```
Comment (Geethree) [inline build.gradle]: HA! I wonder where this object came from.

**Comment by mcslee** [comment]
The demuxing looks fine to me. I'm slightly confused about some of this var0/var1 code and this comment:
 // This loop is unrolled weirdly, but I believe returns a int[] of
    // the indices represented by the fixture

Were you using some kind of tool that was automatically converting code? You might just use this utility for the fixtureToIndices:
https://github.com/heronarts/LX/blob/master/src/heronarts/lx/model/LXFixture.java
LXFixture.Utils.getIndices(LXFixture fixture);

Also I'll just note that you could reduce the amount of redundant code here by just subclassing ArtNetDatagram and overriding the onSend() method.
https://github.com/heronarts/LX/blob/master/src/heronarts/lx/output/ArtNetDatagram.java

But, functionally - this looks like it should do what you want!

**Comment by Geethree** [comment]
Quick not regarding the var0/var1 variables: I believe that is how intellij presents variables of methods from decompiled jars.

**Comment by mcslee** [comment]
Ahh, gotcha. Assuming you are aware, but all the library source is fully available for reference:
https://github.com/heronarts/LX
https://github.com/heronarts/P3LX

**Comment by Geethree** [comment]
Also
http://lx.studio/api/ 

=)

**Comment by meawoppl** [comment]
Thanks, @mcslee. This was a lot of copy pasta, is see the datagram size arg now, I will use that :). Ty!

**Comment by chaqke** [comment]
oh hey, i was on a roadtrip until last night, sorry!  i'll take a look now

**Comment by chaqke** [comment]
looks good, especially with mcslee's suggestions!

**Review by chaqke — APPROVED** [review]
looks good, especially with mcslee's suggestions!

**Comment by meawoppl** [comment]
So @mcslee
- I made the change to turn this into a subclass of `ArtNetDatagram` much smoother.  Thanks!
- I put in a (broken) change to demo what I think you are suggesting on the `LXFixture.Utils.getIndices(fixture)` call.  My IDE and the compiler seem to think that the `public static class Utils` that hangs off the interface `LXFixture` does not resolve from here.  Are we possibly using a different version?  @Geethree ?

I don't quite know enough about Java to say what the behavior should be here, but I am still poking at it.

**Comment by meawoppl** [comment]
image

**Comment by meawoppl** [comment]
Confirmed, @Geethree I think our LX bits are from before when this change landed:
https://github.com/heronarts/LX/commit/01e6a2654c2cd8fc10810a27d0919a0927d64e16#diff-def653c3387776484747ef093daba6dd

Late March or so?

**Comment by Geethree** [comment]
PR to update LX here https://github.com/temple2018/Galaxia/pull/11

**Comment by meawoppl** [comment]
@mcslee I see what you did there:
https://github.com/heronarts/LX/commit/ddb15fc067d324ddec6abef489e1bcdddbbbaaf4

:)

File: src/main/java/org/templegalaxia/DemuxArtNetDatagram.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  public static DemuxArtNetDatagram fromFixture(LXFixture fixture, int universe) {
+    return new DemuxArtNetDatagram(LXFixture.Utils.getIndices(fixture), universe);
+  }
+
+  private DemuxArtNetDatagram(int[] pointIndices, int universe) {
+    super(pointIndices, pointIndices.length, universe);
+    this.pointIndices = pointIndices;
+  }
+
+  public byte luminance(int rgb) {
```
Comment (meawoppl) [inline src/main/java/org/templegalaxia/DemuxArtNetDatagram.java:20]: One question @mcslee , there isn't much documentation on the color model, what are the output ranges of:
`LXColor.`
 - red/green/blue
 - h/s/b
?

Comment (mcslee) [reply src/main/java/org/templegalaxia/DemuxArtNetDatagram.java:20]: The color functions all internally use integers in standard 32-bit ARGB format, one byte each for A, R, G, B. Which is what Processing also uses.

The red() / green() / blue() functions return a byte 0-255.

h() is 0-360, s() and b() are 0-100.

I'll make a note to document those all better.

Source reference:
https://github.com/heronarts/LX/blob/master/src/heronarts/lx/color/LXColor.java#L57

**Review by Geethree — APPROVED** [review]
Is this good to go @meawoppl ?

**Pull request merged**